docs/advanced/customization-guide.md

```diff
@@ -16,8 +15,1007 @@ published: false
 
 ---
 
-***WORK IN PROGRESS.***
+## Overview
+
+NFD provides multiple extension points for vendor and application specific
+labeling:
+
+- [NodeFeatureRule](#nodefeaturerule-custom-resource) objects provide a way to
+  deploy custom labeling rules via the Kubernetes API
+- [*local* feature source](#local-feature-source) of nfd-worker creates
+  labels by executing hooks and reading files
+- [*custom* feature source](#custom-feature-source) of nfd-worker creates
+  labels based on user-specified rules
+
+## NodeFeatureRule custom resource
+
+NodeFeatureRule objects provide an easy way to create vendor or application
+specific labels. It uses a flexible rule-based mechanism for creating labels
+based on node feature.
+
+### A NodeFeatureRule example
+
+Consider the following referential example:
+
+```yaml
+apiVersion: nfd.k8s-sigs.io/v1alpha1
+kind: NodeFeatureRule
+metadata:
+  name: my-rule-object
+spec:
+  rules:
+    - name: "my sgx feature rule"
+      labels:
+        "my-sgx-feature": "true"
+      matchFeatures:
+        - feature: cpu.cpuid
+          matchExpressions:
+            SGX: {op: Exists}
+        - feature: kernel.config
+          matchExpressions:
+            X86_SGX: {op: In, value: ["y"]}
+```
+
+It specifies one rule which creates node label
+`feature.node.kubenernetes.io/my-sgx-feature=true` if both of the following
+conditions are true (`matchFeatures` implements a logical AND over the
+matchers):
+
+- the CPU has SGX capability
+- X86_SGX option has been enabled in the kernel
+
+### NodeFeatureRule controller
+
+NFD-Master acts as the controller for NodeFeatureRule objects. It applies these
+rules on raw feature data received from nfd-worker instances and creates node
+labels, accordingly.
+
+**NOTE** nfd-master is stateless and (re-)labelling only happens when a request
+is received from nfd-worker. That is, in practice rules are evaluated and
+labels for each node are created on intervals specified by the
+[`core.sleepInterval`](worker-configuration-reference#coresleepinterval)
+configuration option (or
+[`-sleep-interval`](worker-commandline-reference#-sleep-interval) command line
+flag) of nfd-worker instances. This means that modification or creation of
+NodeFeatureRule objects does not instantly cause the node labels to be updated.
+Instead, the changes only come visible in node labels as nfd-worker instances
+send their labelling requests.
+
+## Local feature source
+
+NFD-Worker has a special feature source named *local* which is an integration
+point for external feature detectors. It provides a mechanism for pluggable
+extensions, allowing the creation of new user-specific features and even
+overriding built-in labels.
+
+The *local* feature source has two methods for detecting features, hooks and
+feature files.
```

Comment:
some readers might ask can `local` sources info be used with `matchFeatures` with the two other customization features?

Reply:
Good point. Any suggestion how to fix this? *The features discovered by the local source can further be used in label rules of NodeFeatureRule objects and the custom source*(?)

Reply:
The proposal looks good to me!

Reply:
Changed along these lines